streams/src/main/java/org/apache/kafka/streams/processor/internals/PunctuationSchedule.java

```diff
@@ -60,13 +60,17 @@ boolean isCancelled() {
     }
 
     public PunctuationSchedule next(long currTimestamp) {
-        PunctuationSchedule nextSchedule;
-        // we need to special handle the case when it is firstly triggered (i.e. the timestamp
-        // is equal to the interval) by reschedule based on the currTimestamp
-        if (timestamp == 0L)
-            nextSchedule = new PunctuationSchedule(value, currTimestamp + interval, interval, punctuator, cancellable);
+        long nextPunctuationTime;
+        // avoid scheduling a new punctuations immediately, this can happen:
+        // - when using STREAM_TIME punctuation and there was a gap i.e., no data was
+        //   received for more than 2*interval (also happens on first STREAM_TIME punctuation i.e., when timestamp == 0L)
+        // - when using WALL_CLOCK_TIME and there was a gap i.e., punctuation was delayed for more than 2*interval (GC pause, overload, ...)
+        if (timestamp + interval < currTimestamp)
+            nextPunctuationTime = currTimestamp + interval;
```

Comment:
If we do this, future punctuation are not aligned nicely. What I mean is, if I register a schedule of 10 seconds, I want be called
```
schedule:    10, 20, 30, 40, 50, 60
data-ts:     10, 20, 30, 40, 50, 60
punctuation: 10, 20, 30, 40, 50, 60
```
and if we do one punctuation a little late, this should not influence later ones. For example, if "10" gets triggered at 12 I still want to be triggered at 20, 30, but not 22, 32. (I put schedules time in "X" below):
```
schedule:        10, 20, 30, 40, 50, 60 // schedule independent from actual punctuation trigger time!
data-timestamps: 12, 20, 30, 44, 50, 60
punctuations:    12, 20, 30, 44, 50, 60 // ok
```
A bad example would be (with same input data as above):
```
bad schedule:     10, 22, ....40, 54, ....70 // schedule depends on actual punctuation trigger time, and thus gets messed up :(
data-timestamps:  12, 20, 30, 44, 50, 60
bad punctuations: 12, ....30, 44, ....60 // note: after "10" triggered (at 12) we would register "22", and thus 20 would not trigger, but 30 would tigger "22" with ts 30
```
In this example, we only trigger 4 time instead of 6 time as in the "good example" above, because the time shift adds up over time.

We should only "jump" over a punctuation of we have gaps in the data that jumps over a full schedule at once:
```
schedule:        10, 20, 40, 50, 60 // schedule skips "30" and moves to "40" directly
data-timestamps: 12, 34, 40, 50, 60 // record at ts 20 is missing
punctuations:    12, 34, 40, 50, 60 // ok, we "miss" 20 (or trigger it late at 34) and thus get one punctuation less than above---only 5 instead of 6---however, we only have 5 input records instead of 6, thus this is ok
```

Reply:
With my code above, the punctuations stay aligned unless there's a gap bigger than 2*interval. I'll modify the tests to make it clear.

Now in case of gap bigger than 2*interval, it is true that my code above doesn't stay aligned, but I don't see the value of staying aligned. I'll write about that on the [jira issue](https://issues.apache.org/jira/browse/KAFKA-6323).